docs/calendar-draft.md

```diff
@@ -0,0 +1,315 @@
+# Draft Design of Temporal Calendar API
+
+This doc describes a design for first-class support for non-Gregorian [calendars](https://en.wikipedia.org/wiki/Calendar) in Temporal.
+
+## Temporal.Date internal slots
+
+Temporal.Date currently has three internal slots: year, month, and day. (An "internal slot" refers to actual data, as opposed to "properties", which could be computed.)  In this proposal, three more internal slots would be added:
+
+1. calendar
+2. isLeapMonth
+3. era
+
+It is believed that a date in any known calendar can be described using the five-tuple (*year*, *month*, *day*, *isLeapMonth*, *era*).  It is possible that *isLeapMonth* can be merged into *month* by expanding the set of numbers that *month* can represent.
+
+The *calendar* slot contains an object implementing the Temporal.Calendar interface, described below.
+
+## Temporal.Calendar interface
+
+The new Temporal.Calendar interface is a mechanism to allow arbitrary calendar systems to be implemented on top of Temporal.  ***Most users will not encounter the Temporal.Calendar interface directly***, unless they are building or using a non-built-in calendar system.
+
+### Methods on the Temporal.Calendar interface
+
+All of the following methods return new Temporal.Date objects.  All properties of `Temporal.Calendar` are Symbols.
+
+```javascript
+class MyCalendar {
```

Comment:
should this not be made to subclass a base Temporal.Calendar, and thus gain an internal slot?

Reply:
I'll defer on that question.  Is there precedent for making an "interface" in 262 that is intended to be subclassable?

Reply:
I guess it maybe depends on whether being a calendar is a protocol - something any object can participate in - or an identity (which would of course have some methods also).

Reply:
Calendars (like TimeZones) are objects adhering to a protocol

Reply:
Both are surprising to me - the whole point to me of both is nominal cross-realm types that robustly (protocols cant be robust) encapsulate relevant data and operations.

Reply:
They are actual types that can be robustly identified cross realm iff one restricts oneself to the system provided TimeZone/Calendar objects retrieved via string.

However there is the option of providing a custom TimeZone/Calendar object that implement the relevant logic, which is not cross-realmable. This is useful for doing custom stuff, such as:

 * custom timezone around ramadan, where the actual changes aren’t known until very short notice.
 * a hebrew calendar that based on longitude and latitude changes the start/end and length of days.

These can obviously not be cross realmed by default.

Reply:
So then why a protocol, rather than requiring custom ones to subclass the builtin constructors?

Reply:
I proposed using Symbols for the methods (instead of strings) because my understanding is that's the ES2015 way of doing protocols.  I think it's consistent with https://github.com/michaelficarra/proposal-first-class-protocols, for example.  The question of cross-realm is an interesting one. @michaelficarra

Reply:
That proposal will need to cover string methods as well to be able to describe thenable, toStringable, valueOfable, etc, but I agree symbols are appropriate for protocols.

However, I don't think TimeZone or Calendar are things that should be protocol-based.

Reply:
I added this to the doc:

> An open question is whether this "interface" should be a protocol or an identity.  See issue #289.  The following section assumes it is a protocol.

I am resolving this thread because this is not the best place to have this discussion.  I'd like to check in this PR with open questions documented.